exporters/otlp/otlphttp/driver.go

```diff
@@ -0,0 +1,269 @@
+		cfg.maxAttempts = DefaultMaxAttempts
+	}
+	if cfg.maxAttempts > DefaultMaxAttempts {
+		cfg.maxAttempts = DefaultMaxAttempts
+	}
+	if cfg.backoff <= 0 {
+		cfg.backoff = DefaultBackoff
+	}
+	client := &http.Client{}
+	if cfg.tlsCfg != nil {
+		transport := defaultTransportClone.Clone()
+		transport.TLSClientConfig = cfg.tlsCfg
+		client.Transport = transport
+	}
+	return &driver{
+		client: client,
+		cfg:    cfg,
+		stopCh: make(chan struct{}),
+	}
+}
+
+// Start implements otlp.ProtocolDriver.
+func (d *driver) Start(ctx context.Context) error {
+	// nothing to do
+	return nil
+}
+
+// Stop implements otlp.ProtocolDriver.
+func (d *driver) Stop(ctx context.Context) error {
+	close(d.stopCh)
+	return nil
+}
+
+// ExportMetrics implements otlp.ProtocolDriver.
+func (d *driver) ExportMetrics(ctx context.Context, cps metricsdk.CheckpointSet, selector metricsdk.ExportKindSelector) error {
+	rms, err := transform.CheckpointSet(ctx, selector, cps, 1)
+	if err != nil {
+		return err
+	}
+	if len(rms) == 0 {
+		return nil
+	}
+	pbRequest := &colmetricspb.ExportMetricsServiceRequest{
+		ResourceMetrics: rms,
+	}
+	rawRequest, err := pbRequest.Marshal()
+	if err != nil {
+		return err
+	}
+	return d.send(ctx, rawRequest, d.cfg.metricsURLPath)
+}
+
+// ExportTraces implements otlp.ProtocolDriver.
+func (d *driver) ExportTraces(ctx context.Context, ss []*tracesdk.SpanSnapshot) error {
+	protoSpans := transform.SpanData(ss)
+	if len(protoSpans) == 0 {
+		return nil
+	}
+	pbRequest := &coltracepb.ExportTraceServiceRequest{
+		ResourceSpans: protoSpans,
+	}
+	rawRequest, err := pbRequest.Marshal()
+	if err != nil {
+		return err
+	}
+	return d.send(ctx, rawRequest, d.cfg.tracesURLPath)
+}
+
+func (d *driver) send(ctx context.Context, rawRequest []byte, urlPath string) error {
+	address := fmt.Sprintf("%s://%s%s", d.getScheme(), d.cfg.address, urlPath)
+	var cancel context.CancelFunc
+	ctx, cancel = d.contextWithStop(ctx)
+	defer cancel()
+	for i := 0; i < d.cfg.maxAttempts; i++ {
+		response, err := d.singleSend(ctx, rawRequest, address)
+		if err != nil {
+			return err
+		}
+		// We don't care about the body, so close it
+		// immediately.
+		response.Body.Close()
+		switch response.StatusCode {
+		case http.StatusOK:
+			return nil
+		case http.StatusTooManyRequests:
+			fallthrough
+		case http.StatusServiceUnavailable:
+			select {
+			case <-time.After(getWaitDuration(d.cfg.backoff, i)):
+				continue
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		default:
+			return fmt.Errorf("Failed with HTTP status %s", response.Status)
+		}
+	}
+	return fmt.Errorf("Failed to send data to %s after %d tries", address, d.cfg.maxAttempts)
+}
+
+func (d *driver) getScheme() string {
+	if d.cfg.insecure {
+		return "http"
+	}
+	return "https"
+}
+
+func getWaitDuration(backoff time.Duration, i int) time.Duration {
+	// Strategy: after nth failed attempt, attempt resending after
+	// k * initialBackoff, where k is a random number in range [0,
+	// 2^n-1).
```

Comment:
This seems like it could lead to inconsistent behavior and potentially little to no backoff at all.  Would it be better to do straight exponential backoff with random jitter to avoid stampeding?  

```go
exp := (int64)1 << (i+1);
jit := rand.Int63n(5)-2 // [-2,3)
res := backoff * time.Duration(exp) + jit
if res <=0 {
    res = backoff
}
return res
```

Reply:
I have taken this algo from https://en.wikipedia.org/wiki/Exponential_backoff#Example_exponential_backoff_algorithm. Adding some jitter could be a good idea, though.

Reply:
Would you put a link in the code, please?

Reply:
Done.